pkg/ingest/ingest.go

```diff
@@ -0,0 +1,101 @@
+package ingest
+
+import (
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/thanos-io/thanos/pkg/store/storepb"
+
+	"github.com/thanos-community/obslytics/pkg/dataframe"
+	"github.com/thanos-community/obslytics/pkg/output"
+)
+
+type Ingestor interface {
+	Ingest(*storepb.Series) error
+}
+
+// Options for a single aggregation.
+type AggrOption struct {
+	// Should the aggregation be used?
+	Enabled bool
+	// Column to store the aggregation at
+	Column string
+}
+
+// Collections of aggregations-related options. Determines what aggregations are enabled etc..
+type aggrsOptions struct {
+	// If true (default), the samples should repeat the same time for every day.
+	// This puts additional requirement for 24h to be divisible by the resolution.
+	MidnightAligned bool
+
+	Sum   AggrOption
+	Count AggrOption
+	Min   AggrOption
+	Max   AggrOption
+}
+
+// By default, all aggregations are disabled and target columns set with `_` prefix.
+func NewAggregationsOptions() (ao aggrsOptions) {
+	ao.MidnightAligned = true
+
+	ao.Sum.Column = "_sum"
+	ao.Count.Column = "_count"
+	ao.Min.Column = "_min"
+	ao.Max.Column = "_max"
+	return ao
+}
+
+// Ingests the data and produces aggregations at the specified resolution.
+type AggregatedIngestor struct {
+	Resolution  time.Duration
+	aggrOptions aggrsOptions
+}
+
+// Ingest the data to an aggregated set.
+func (i AggregatedIngestor) Ingest(s *storepb.Series) error {
+	// TODO(inecas): implement
+	return nil
+}
+
+// Return already aggregated data from previous samples while cleaning
+// the buffer.
+func (i AggregatedIngestor) Flush() dataframe.Dataframe {
```

Comment:
Is there anything to flush really?

Reply:
We just want to return result, no?

Anyway, not big deal let's not overdesign this, let's see what would easier to user later on

Reply:
I expect the aggregator would need to keep some internal state before flushing happens: I take this as atomic operation that returns whatever was accumulated so far and clears the buffer at the same time.

I'm ok with revisiting once full example is there.

Reply:
More code there now to better understand the end2end